## Libraries

In [1]:
import pandas as pd

## Load Data

In [2]:
raw_data = pd.read_csv('3.1-data-sheet-guvi-courses.csv')

In [3]:
raw_data

course_id                                       course_title  \
0       41295.0               Learn HTML5 Programming From Scratch   
1       59014.0                     Coding for Entrepreneurs Basic   
2      625204.0                         The Web Developer Bootcamp   
3      173548.0  Build Your First Website in 1 Week with HTML5 ...   
4      764164.0              The Complete Web Developer Course 2.0   
...         ...                                                ...   
3675  1216340.0                  Learn to Play Sweet Child of Mine   
3676  1026762.0  Coaching Course: Rhythm Instruction for Music ...   
3677  1223240.0  Learn Pirates of the Caribbean by Ear on the P...   
3678  1224186.0  Learn to Play Piano Like a Pro - Easy Piano Co...   
3679  1234656.0                              4 Week Rhythm Mastery   

                                                    url  price  \
0     https://www.udemy.com/learn-html5-programming-...    0.0   
1     https://www.udemy.com/coding-for-entrepreneurs...    0.0   
2     https://www.udemy.com/the-web-developer-bootcamp/  200.0   
3     https://www.udemy.com/build-your-first-website...    0.0   
4     https://www.udemy.com/the-complete-web-develop...  200.0   
...                                                 ...    ...   
3675  https://www.udemy.com/learn-to-play-sweet-chil...   20.0   
3676  https://www.udemy.com/coaching-bundle-rhythm-i...  125.0   
3677  https://www.udemy.com/learn-pirates-of-the-car...   20.0   
3678  https://www.udemy.com/learn-to-play-piano-like...   35.0   
3679       https://www.udemy.com/4-week-rhythm-mastery/   20.0   

      num_subscribers  num_reviews  num_lectures           level  Rating  \
0            268923.0       8629.0          45.0  Beginner Level    0.82   
1            161029.0        279.0          27.0    Expert Level    0.69   
2            121584.0      27445.0         342.0  Beginner Level    0.89   
3            120291.0       5924.0          30.0      All Levels    0.78   
4            114512.0      22412.0         304.0  Beginner Level    0.55   
...               ...          ...           ...             ...     ...   
3675              0.0          0.0          23.0      All Levels    0.31   
3676              0.0          0.0           7.0  Beginner Level    0.06   
3677              0.0          0.0           6.0      All Levels    0.08   
3678              0.0          0.0          13.0  Beginner Level    0.78   
3679              0.0          0.0          28.0      All Levels    0.01   

      content_duration   published_timestamp                   subject  
0            10.500000  2013-02-14T07:03:41Z  Subject: Web Development  
1             3.500000  2013-06-09T15:51:55Z  Subject: Web Development  
2            43.000000  2015-11-02T21:13:27Z  Subject: Web Development  
3             3.000000  2014-04-08T16:21:30Z  Subject: Web Development  
4            30.500000  2016-03-08T22:28:36Z  Subject: Web Development  
...                ...                   ...                       ...  
3675          1.500000  2017-06-30T17:33:29Z       Musical Instruments  
3676          0.633333  2017-03-10T17:06:46Z       Musical Instruments  
3677          0.533333  2017-05-22T17:14:43Z       Musical Instruments  
3678          0.516667  2017-05-30T00:19:02Z       Musical Instruments  
3679          2.500000  2017-07-05T16:07:44Z       Musical Instruments  

[3680 rows x 12 columns]

In [4]:
raw_data.isnull().sum()

course_id              4
course_title           4
url                    4
price                  4
num_subscribers        4
num_reviews            4
num_lectures           4
level                  4
Rating                 4
content_duration       4
published_timestamp    4
subject                3
dtype: int64

## Data PreProcessing

In [15]:
data = raw_data.dropna()

In [16]:
data.isnull().sum()

course_id              0
course_title           0
url                    0
price                  0
num_subscribers        0
num_reviews            0
num_lectures           0
level                  0
Rating                 0
content_duration       0
published_timestamp    0
subject                0
dtype: int64

In [17]:
data.describe(include= 'all')

course_id                                       course_title  \
count   3.676000e+03                                               3676   
unique           NaN                                               3663   
top              NaN  Creating an animated greeting card via Google ...   
freq             NaN                                                  3   
mean    6.757535e+05                                                NaN   
std     3.431304e+05                                                NaN   
min     8.324000e+03                                                NaN   
25%     4.074740e+05                                                NaN   
50%     6.876920e+05                                                NaN   
75%     9.608140e+05                                                NaN   
max     1.282064e+06                                                NaN   

                                                      url        price  \
count                                                3676  3676.000000   
unique                                               3672          NaN   
top     https://www.udemy.com/make-your-own-business-w...          NaN   
freq                                                    2          NaN   
mean                                                  NaN    66.115343   
std                                                   NaN    61.056073   
min                                                   NaN     0.000000   
25%                                                   NaN    20.000000   
50%                                                   NaN    45.000000   
75%                                                   NaN    95.000000   
max                                                   NaN   200.000000   

        num_subscribers   num_reviews  num_lectures       level       Rating  \
count       3676.000000   3676.000000   3676.000000        3676  3676.000000   
unique              NaN           NaN           NaN           4          NaN   
top                 NaN           NaN           NaN  All Levels          NaN   
freq                NaN           NaN           NaN        1925          NaN   
mean        3199.260881    156.309848     40.129761         NaN     0.610868   
std         9486.582966    935.674518     50.398507         NaN     0.334287   
min            0.000000      0.000000      0.000000         NaN     0.000000   
25%          112.000000      4.000000     15.000000         NaN     0.280000   
50%          912.500000     18.000000     25.000000         NaN     0.760000   
75%         2558.000000     67.000000     46.000000         NaN     0.930000   
max       268923.000000  27445.000000    779.000000         NaN     1.000000   

        content_duration   published_timestamp                   subject  
count        3676.000000                  3676                      3676  
unique               NaN                  3672                         4  
top                  NaN  2014-06-23T16:03:41Z  Subject: Web Development  
freq                 NaN                     2                      1203  
mean            4.096137                   NaN                       NaN  
std             6.054948                   NaN                       NaN  
min             0.000000                   NaN                       NaN  
25%             1.000000                   NaN                       NaN  
50%             2.000000                   NaN                       NaN  
75%             4.500000                   NaN                       NaN  
max            78.500000                   NaN                       NaN

In [18]:
df_encoded = pd.get_dummies(data, columns=['level', 'subject'], drop_first=True, dtype=int)

In [19]:
df_encoded.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3676 entries, 0 to 3679
Data columns (total 16 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   course_id                         3676 non-null   float64
 1   course_title                      3676 non-null   object 
 2   url                               3676 non-null   object 
 3   price                             3676 non-null   float64
 4   num_subscribers                   3676 non-null   float64
 5   num_reviews                       3676 non-null   float64
 6   num_lectures                      3676 non-null   float64
 7   Rating                            3676 non-null   float64
 8   content_duration                  3676 non-null   float64
 9   published_timestamp               3676 non-null   object 
 10  level_Beginner Level              3676 non-null   int64  
 11  level_Expert Level                3676 non-null   int64  
 12  level_Inter

## Feature Selection

In [20]:
data_selection = df_encoded.drop(['course_id','course_title','url','content_duration','published_timestamp'], axis =1)

In [22]:
data_selection.describe()

price  num_subscribers   num_reviews  num_lectures       Rating  \
count  3676.000000      3676.000000   3676.000000   3676.000000  3676.000000   
mean     66.115343      3199.260881    156.309848     40.129761     0.610868   
std      61.056073      9486.582966    935.674518     50.398507     0.334287   
min       0.000000         0.000000      0.000000      0.000000     0.000000   
25%      20.000000       112.000000      4.000000     15.000000     0.280000   
50%      45.000000       912.500000     18.000000     25.000000     0.760000   
75%      95.000000      2558.000000     67.000000     46.000000     0.930000   
max     200.000000    268923.000000  27445.000000    779.000000     1.000000   

       level_Beginner Level  level_Expert Level  level_Intermediate Level  \
count           3676.000000         3676.000000               3676.000000   
mean               0.345756            0.015778                  0.114799   
std                0.475679            0.124633                  0.318822   
min                0.000000            0.000000                  0.000000   
25%                0.000000            0.000000                  0.000000   
50%                0.000000            0.000000                  0.000000   
75%                1.000000            0.000000                  0.000000   
max                1.000000            1.000000                  1.000000   

       subject_Graphic Design  subject_Musical Instruments  \
count             3676.000000                  3676.000000   
mean                 0.163765                     0.184984   
std                  0.370113                     0.388337   
min                  0.000000                     0.000000   
25%                  0.000000                     0.000000   
50%                  0.000000                     0.000000   
75%                  0.000000                     0.000000   
max                  1.000000                     1.000000   

       subject_Subject: Web Development  
count                       3676.000000  
mean                           0.327258  
std                            0.469276  
min                            0.000000  
25%                            0.000000  
50%                            0.000000  
75%                            1.000000  
max                            1.000000

## Spliting the Data

In [23]:
X = data_selection.drop('Rating', axis = 1)
y = data_selection['Rating']

In [49]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


### Finding the best Regression

In [50]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

# Define a dictionary of regression models
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(),
    "Random Forest": RandomForestRegressor(),
    "SVR": SVR(),
}

# Initialize dictionaries to store evaluation metrics
mse_scores = {}
rmse_scores = {}
mae_scores = {}
r2_scores = {}

# Iterate through the models and calculate metrics
for model_name, model in models.items():
    model.fit(X_train, y_train)  # Train the model
    y_pred = model.predict(X_test)  # Make predictions

    # Calculate evaluation metrics
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    # Store the metrics in the dictionaries
    mse_scores[model_name] = mse
    rmse_scores[model_name] = rmse
    mae_scores[model_name] = mae
    r2_scores[model_name] = r2

# Print the evaluation metrics for each model
for model_name in models.keys():
    print(f"Model: {model_name}")
    print(f"Mean Squared Error (MSE): {mse_scores[model_name]:.2f}")
    print(f"Root Mean Squared Error (RMSE): {rmse_scores[model_name]:.2f}")
    print(f"Mean Absolute Error (MAE): {mae_scores[model_name]:.2f}")
    print(f"R-squared (R2) Score: {r2_scores[model_name]:.2f}")
    print("=" * 50)


Model: Linear Regression
Mean Squared Error (MSE): 0.09
Root Mean Squared Error (RMSE): 0.30
Mean Absolute Error (MAE): 0.26
R-squared (R2) Score: 0.18
Model: Decision Tree
Mean Squared Error (MSE): 0.13
Root Mean Squared Error (RMSE): 0.36
Mean Absolute Error (MAE): 0.24
R-squared (R2) Score: -0.18
Model: Random Forest
Mean Squared Error (MSE): 0.07
Root Mean Squared Error (RMSE): 0.27
Mean Absolute Error (MAE): 0.20
R-squared (R2) Score: 0.36
Model: SVR
Mean Squared Error (MSE): 0.12
Root Mean Squared Error (RMSE): 0.35
Mean Absolute Error (MAE): 0.29
R-squared (R2) Score: -0.13


#### Hyperparameter Tuning 

In [51]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import GridSearchCV

# Define the hyperparameter grid
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Create the Random Forest regressor
rf = RandomForestRegressor(random_state=42)

# Create the GridSearchCV object
grid_search = GridSearchCV(estimator=rf, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)

# Fit the grid search to your data
grid_search.fit(X_train, y_train)

# Get the best hyperparameters
best_params = grid_search.best_params_
best_rf = grid_search.best_estimator_


### Testing and Predicting

In [52]:
# Use the best Random Forest model to make predictions on the test data
y_pred = best_rf.predict(X_test)

# Calculate evaluation metrics on the test predictions
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import numpy as np

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"R-squared (R2) Score: {r2:.2f}")

Mean Squared Error (MSE): 0.07
Root Mean Squared Error (RMSE): 0.26
Mean Absolute Error (MAE): 0.20
R-squared (R2) Score: 0.39


### Interpretation

In [53]:
# Assuming 'best_rf' is your tuned Random Forest model
importances = best_rf.feature_importances_

# Create a DataFrame to associate feature names with their importances
feature_importance_df = pd.DataFrame({'Feature': X.columns, 'Importance': importances})

# Sort the DataFrame by importance in descending order
feature_importance_df = feature_importance_df.sort_values(by='Importance', ascending=False)

# Display the top N most important features
top_n = 20  # You can change this value to see more or fewer top features
print(f"Top {top_n} most important features:")
print(feature_importance_df.head(top_n))


Top 20 most important features:
                            Feature  Importance
1                   num_subscribers    0.413337
8       subject_Musical Instruments    0.240311
3                      num_lectures    0.107674
2                       num_reviews    0.102997
0                             price    0.065724
7            subject_Graphic Design    0.026969
9  subject_Subject: Web Development    0.022676
4              level_Beginner Level    0.011910
6          level_Intermediate Level    0.007276
5                level_Expert Level    0.001127


1. **`num_subscribers`** (Importance: 0.413337):
   - This feature represents the number of subscribers for the course.
   - Importance: The most critical feature, indicating that the number of subscribers has a substantial impact on course ratings. Courses with more subscribers tend to receive higher ratings.

2. **`subject_Musical Instruments`** (Importance: 0.240311):
   - This is a binary (0/1) feature indicating whether the course falls under the "Musical Instruments" subject.
   - Importance: Courses related to musical instruments have significant importance in predicting course ratings.

3. **`num_lectures`** (Importance: 0.107674):
   - This feature represents the number of lectures in the course.
   - Importance: The number of lectures is another influential factor. Courses with more lectures may be perceived as more comprehensive and receive higher ratings.

4. **`num_reviews`** (Importance: 0.102997):
   - This feature denotes the number of reviews for the course.
   - Importance: The number of reviews is essential. Courses with more reviews may be considered more credible and trustworthy, affecting their ratings.

5. **`price`** (Importance: 0.065724):
   - This feature represents the price of the course.
   - Importance: The course price also has a significant impact on ratings. Lower-priced courses might attract more subscribers and, in turn, receive higher ratings.

6. **`subject_Graphic Design`** (Importance: 0.026969):
   - This is a binary (0/1) feature indicating whether the course falls under the "Graphic Design" subject.
   - Importance: Courses related to graphic design contribute to predicting ratings but are less influential than other subjects.

7. **`subject_Subject: Web Development`** (Importance: 0.022676):
   - This binary feature indicates whether the course falls under the "Web Development" subject.
   - Importance: Web development courses also have some importance in predicting ratings.

8. **`level_Beginner Level`** (Importance: 0.011910):
   - This binary feature indicates whether the course is categorized as "Beginner Level."
   - Importance: Courses targeted at beginners are a factor but have less influence compared to other levels.

9. **`level_Intermediate Level`** (Importance: 0.007276):
   - This binary feature denotes whether the course is categorized as "Intermediate Level."
   - Importance: Intermediate-level courses also contribute but with relatively lower importance.

10. **`level_Expert Level`** (Importance: 0.001127):
   - This binary feature indicates whether the course is categorized as "Expert Level."
   - Importance: Expert-level courses have the least influence on predicting ratings among the factors considered.

In summary, these features provide insights into the factors that influence course ratings. Features related to the course's popularity (number of subscribers, reviews), content (number of lectures), price, and subject matter play significant roles in determining course ratings.

In [40]:
import joblib

# Save the model to a file
joblib.dump(best_rf, 'random_forest_model.pkl')


['random_forest_model.pkl']